src/internal/snapshotdb/snapshot_crud.go

```diff
@@ -0,0 +1,67 @@
+package snapshotdb
+
+import (
+	"context"
+	"github.com/jmoiron/sqlx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset"
+	"github.com/pachyderm/pachyderm/v2/src/version"
+)
+
+const (
+	selectSnapshotPrefix = `SELECT * FROM recovery.snapshots`
+	insertSnapshot       = `
+		insert into recovery.snapshots (chunkset_id, pachyderm_version, metadata) 
+		values ($1, $2, $3) returning id`
+	insertSnapshotWithoutMetadata = `
+		insert into recovery.snapshots (chunkset_id, pachyderm_version) 
+		values ($1, $2) returning id`
+)
+
+func CreateSnapshot(ctx context.Context, tx *pachsql.Tx, s *fileset.Storage, metadata []byte) (SnapshotID, error) {
+	chunksetID, err := s.CreateChunkSet(ctx, tx)
+	if err != nil {
+		return 0, errors.Wrap(err, "create chunkset")
+	}
+	var id SnapshotID
+	if len(metadata) == 0 {
```

Comment:
We have `pgjsontypes.StringMap` for reading/storing map[string]string values into postgres.  This maps empty maps to `{}` as the CHECK constraint on the column requires.